hooks/src/index.d.ts

```diff
@@ -62,7 +62,7 @@ export function useRef<T>(initialValue: null): RefObject<T>;
 export function useRef<T>(initialValue: T): Ref<T>;
 export function useRef<T>(): Ref<T | undefined>;
 
-type EffectCallback = () => void | (() => void);
+type EffectCallback = () => void | Promise<void> | (() => void);
```

Comment:
It looks like React doesn't allow for this https://www.runpkg.com/?@types/react@17.0.21/index.d.ts#901

Reply:
I didn't think it did, though would more lax types from Preact be an issue?

Reply:
I see what we're aiming to do, the thing that scares me about this is that we "allow" people to invoke race conditions, what if this effect invokes twice due to two renders with two different sets of arguments making this downloadData return two distinct values. What's stopping the user from going into erroneous states? It's implied that the user uses an `AbortController`, ... but it's hard to allow this and then not extensively document what they need to check for.

Reply:
Definitely.

Certainly happy to close this out if you're of the opinion that it's likely to do more harm than good.

I have no arguments for or against, just was scrolling through open issues and came across Jason's comment asking for a PR.

Reply:
I mean, that's only my opinion, open to anyone else chiming in, just thought I'd leave my reasoning 😄

Reply:
Agree with @JoviDeCroock . In most client projects people tend to want to add async/await and introduce the mentioned race conditions unknowingly.